okhttp/src/main/java/okhttp3/Dispatcher.kt

```diff
@@ -0,0 +1,234 @@
+      if (existingCall.host() == host) return existingCall
+    }
+    return null
+  }
+
+  /**
+   * Cancel all calls currently enqueued or executing. Includes calls executed both
+   * [synchronously][Call.execute] and [asynchronously][Call.enqueue].
+   */
+  @Synchronized fun cancelAll() {
+    for (call in readyAsyncCalls) {
+      call.get().cancel()
+    }
+    for (call in runningAsyncCalls) {
+      call.get().cancel()
+    }
+    for (call in runningSyncCalls) {
+      call.cancel()
+    }
+  }
+
+  /**
+   * Promotes eligible calls from [readyAsyncCalls] to [runningAsyncCalls] and runs them on the
+   * executor service. Must not be called with synchronization because executing calls can call
+   * into user code.
+   *
+   * @return true if the dispatcher is currently running calls.
+   */
+  private fun promoteAndExecute(): Boolean {
+    assert(!Thread.holdsLock(this))
+
+    val executableCalls = ArrayList<AsyncCall>()
+    val isRunning: Boolean
+    synchronized(this) {
+      val i = readyAsyncCalls.iterator()
+      while (i.hasNext()) {
+        val asyncCall = i.next()
+
+        if (runningAsyncCalls.size >= this.maxRequests) break // Max capacity.
+        if (asyncCall.callsPerHost().get() >= this.maxRequestsPerHost) continue // Host max capacity.
+
+        i.remove()
+        asyncCall.callsPerHost().incrementAndGet()
+        executableCalls.add(asyncCall)
+        runningAsyncCalls.add(asyncCall)
+      }
+      isRunning = runningCallsCount() > 0
+    }
+
+    for (i in 0 until executableCalls.size) {
+      val asyncCall = executableCalls[i]
+      asyncCall.executeOn(executorService())
+    }
+
+    return isRunning
+  }
+
+  /** Used by `Call#execute` to signal it is in-flight.  */
+  @Synchronized internal fun executed(call: RealCall) {
+    runningSyncCalls.add(call)
+  }
+
+  /** Used by `AsyncCall#run` to signal completion.  */
+  fun finished(call: AsyncCall) {
+    call.callsPerHost().decrementAndGet()
+    finished(runningAsyncCalls, call)
+  }
+
+  /** Used by `Call#execute` to signal completion.  */
+  internal fun finished(call: RealCall) {
+    finished(runningSyncCalls, call)
+  }
+
+  private fun <T> finished(calls: Deque<T>, call: T) {
+    val idleCallback: Runnable?
+    synchronized(this) {
+      if (!calls.remove(call)) throw AssertionError("Call wasn't in-flight!")
+      idleCallback = this.idleCallback
+    }
+
+    val isRunning = promoteAndExecute()
+
+    if (!isRunning && idleCallback != null) {
+      idleCallback.run()
+    }
+  }
+
+  /** Returns a snapshot of the calls currently awaiting execution.  */
+  @Synchronized fun queuedCalls(): List<Call> {
+    return Collections.unmodifiableList(readyAsyncCalls.map { it.get() })
+  }
+
+  /** Returns a snapshot of the calls currently being executed.  */
+  @Synchronized fun runningCalls(): List<Call> {
+    return Collections.unmodifiableList(runningSyncCalls + runningAsyncCalls.map { it.get() })
```

Comment:
method reference? might be longer but I kind of remember the method counts to be lower?

Reply:
The code gets inlined so there should be no effect on method count. Also, bytecode for both versions looks pretty similar.